README.md
## 版本
v0.1.1

## 背景
PyQt在进行任何耗时操作时将堵塞Qt的渲染导致Qt界面卡顿，Qt程序免不了耗时操作，故设计此工具用简单的操作注册运行耗时任务，并完成任务和渲染的通信/交互。

## 说明
本repo用于PyQt耗时任务执行/操作
- By: 大宝天天见丶(zzmx-sudo)

基于PyQt5开发

所用技术栈：

- PyQt5

依赖:

- PyQt5
- psutil

## 快速入门
```python
import sys
import os
from PyQt5.QtWidgets import QMainWindow, QApplication, QPushButton, QTextEdit
from pyqt_ipc import CreateIPC

# 首先有个简单的Qt案例
class Example(QMainWindow):

    def __init__(self):
        super(Example, self).__init__()

        self.initUi()

    def initUi(self):
        # 创建任务ipc和渲染ipc对象
        self.ipcMain, self.ipcReanderer = CreateIPC(self)

        self.setWindowTitle("PyQt-IPC Demo")
        self.resize(400, 400)
        self.btn = QPushButton("点我展示文本内容", self)
        self.btn.resize(self.btn.sizeHint())
        self.btn.move(10, 10)
        # 控件绑定 启动任务事件
        self.btn.clicked.connect(lambda :self.ipcMain.start("read_file", "./pyqt_ipc/demo_aaa.txt"))

        self.textBox = QTextEdit(self)
        self.textBox.resize(380, 350)
        self.textBox.move(10, 40)

        self._setting_async_tasks()

        self.show()
    
    def _setting_async_tasks(self):
        # 注册任务
        self.ipcMain.registry("read_file", self._read_file)
        # 定义监听任务结果回调,允许单任务多个监听,多次调用即可
        self.ipcReanderer.on("read_file", self.textBox.append)
        # 单次监听任务结果回调,同样允许单任务多个监听
        self.ipcReanderer.once("read_file", self.textBox.setText)

    @staticmethod
    def _read_file(file_path):
        """定义任务函数"""
        if not os.path.exists(file_path):
            return ""

        with open(file_path, "r") as f:
            return f.read()


if __name__ == '__main__':
    app = QApplication(sys.argv)
    window = Example()
    sys.exit(app.exec_())
```

#### 还可以是无限循环进行下去的任务
```python
# 先导入任务迭代器，迭代器默认任务无限循环运行
from pyqt_ipc.task import TaskIterator

# 将你的任务用TaskIterator包一层即可,若是有限次数任务,请传入cycles数
# 需要注意任务返回结果频率,若是瞬时(就像示例中那样)且频繁的任务,请适当添加阻塞,否则Qt将因吃不消导致崩溃
self.ipcMain.registry("read_file", TaskIterator(self._read_file))

# 当你想关闭无限任务时，调用一次cancel即可
self.ipcMain.cancel("read_file")
```


CHANGELOG.md
# PyQt-IPC change log

## [0.1.0] - 2023-7-11

* 0.1.0版本发布

## [0.1.1] - 2023-7-16

### 新增：
- 加入开启任务时参数合法性校验

### 修复：
- 修复因任务为TaskIterator实例对象时的闪退

pyqt_ipc/main.py
import time
import inspect
from multiprocessing import Process, Queue

import psutil
from PyQt5.QtCore import pyqtSignal, QThread

from .task import TaskManager, TaskIterator
from .exception import ProcessException, OperationException


__all__ = [
    "IPCMain"
]

task_q = Queue()
result_q = Queue()

class WatchThread(QThread):
    """
    Watch结果线程
    """
    signal = pyqtSignal(tuple)

    def __init__(self, run_flag=True):
        super(WatchThread, self).__init__()
        self._run_flag = run_flag

    def run(self):

        while True:
            if not self._run_flag:
                break

            if result_q.empty():
                time.sleep(0.2)
                continue

            result = result_q.get()
            self.signal.emit(result)

    @property
    def run_flag(self):

        return self._run_flag

    @run_flag.setter
    def run_flag(self, newValue):

        self._run_flag = newValue

class IPCMain:

    def __init__(self, window):

        self._window = window
        self._task_datasets = {}
        self._proc = None
        self._watch_thread = None
        self._listen_always_tasks = {}
        self._listen_once_tasks = {}

    def registry(self, task_name, task):
        """
        注册/更新任务
        :param task_name: 任务名称
        :param task: 任务实体
        :return: None
        """
        if task_name in self._task_datasets:
            task_q.put(("modify", task_name, task))
        else:
            task_q.put(("add", task_name, task))

        self._task_datasets.update({task_name: task})

    def start(self, task_name, *args, **kwargs):
        """
        启动任务
        :param task_name: 任务名称
        :param args: 执行任务传递的位置参数
        :param kwargs: 执行任务传递的关键字参数
        :return: None
        """
        if task_name not in self._task_datasets:
            raise ProcessException("请先完成任务注册后再启动")

        if not self._validate_task_params(task_name, *args, **kwargs):
            raise OperationException("传递给任务的参数不合法")

        task_q.put(("start", task_name, args, kwargs))

    def cancel(self, task_name):
        """
        停止任务
        :param task_name: 任务名称
        :return: None
        """
        if task_name not in self._task_datasets:
            raise ProcessException("任务完成注册并运行后才有停止的可能")

        task_q.put(("stop", task_name))

    def bind_quit(self):
        """
        给Qt退出增加IPC退出操作
        :return: None
        """
        closeEvent = self._window.closeEvent

        def _quit(event):
            if self._watch_thread is not None:
                self._watch_thread.run_flag = False
                self._watch_thread.quit()

            task_q.put(("stop-all",))
            self.__kill_proc()

            closeEvent(event)

        self._window.closeEvent = _quit

    def run(self):
        """
        任务管理进程启动
        :return: None
        """
        self._proc = Process(target=TaskManager.run_ever, args=(task_q, result_q))
        self._proc.daemon = True
        self._proc.start()

        self._watch_thread = WatchThread(True)
        self._watch_thread.signal.connect(self._callback)
        self._watch_thread.start()

    def _callback(self, result):
        """
        结果回调
        :param result: 任务完成后生成的结果
        :return: None
        """
        task_name = result[0]
        if task_name in self._listen_always_tasks:
            self._listen_always_tasks[task_name](*result[1:])

        if task_name in self._listen_once_tasks:
            self._listen_once_tasks[task_name](*result[1:])
            del self._listen_once_tasks[task_name]

    def _validate_task_params(self, task_name, *args, **kwargs):
        """
        校验传递给任务参数是否合法
        :param task_name: 任务名称
        :param args: 传递给任务的位置参数
        :param kwargs: 传递给任务的关键字参数
        :return: bool -> 是否校验成功
        """
        task = self._task_datasets[task_name]
        if isinstance(task, TaskIterator):
            task = task.taskProto
        sig = inspect.signature(task)
        params = sig.parameters
        if len(args) > len(params):
            return False

        kwargs_keys = [x for x in params][len(args):]
        for kwargs_key in kwargs_keys:
            if params[kwargs_key].default is inspect._empty and kwargs_key not in kwargs:
                return False

        return True

    @property
    def listen_always_tasks(self):

        return self._listen_always_tasks

    @listen_always_tasks.setter
    def listen_always_tasks(self, newValue):

        self._listen_always_tasks = newValue

    @property
    def listen_once_tasks(self):

        return self._listen_once_tasks

    @listen_once_tasks.setter
    def listen_once_tasks(self, newValue):

        self._listen_once_tasks = newValue

    def __kill_proc(self):
        """
        任务管理进程关闭
        :return: None
        """
        if self._proc is None:
            return

        process = psutil.Process(self._proc.pid)
        for child in process.children(recursive=True):
            child.kill()

        process.kill()


pyqt_ipc/task.py
import time
from threading import Thread

from .exception import RegistryException, OperationException


__all__ = [
    "TaskIterator",
    "TaskManager"
]

class TaskIterator:

    def __init__(self, task, cycles=0):
        self.__tasks = task
        self._cycles = cycles
        self.__cycle = 0

    def __iter__(self):

        return self

    def __next__(self):

        if self.__cycle < self._cycles or self._cycles == 0:
            self.__cycle += 1
            return self.__tasks
        else:
            raise StopIteration

    def reload(self):
        """
        重置任务
        :return: None
        """
        self.__cycle = 0

    @property
    def taskProto(self):

        return self.__tasks

    @taskProto.setter
    def taskProto(self, newValue):

        raise OperationException("不支持通过该方式修改任务本体")

class Task:

    def __init__(self, task_name, task, result_q):
        if isinstance(task, TaskIterator):
            self._task = task
        elif hasattr(task, "__call__"):
            self._task = TaskIterator(task, cycles=1)
        else:
            raise RegistryException("任务必须为可调用的,它可以是一个函数或TaskIterator实例对象")

        self._run = False
        self._task_name = task_name
        self._result_q = result_q

    def run(self, *args, **kwargs):
        """
        启动任务
        :return: None
        """
        self._run = True
        for task in self._task:
            if not self._run:
                return

            result = task(*args, **kwargs)
            if isinstance(result, tuple):
                self._result_q.put((self._task_name, *result))
            else:
                self._result_q.put((self._task_name, result))

        self._run = False

    def quit(self):
        self._run = False

    def reload(self):
        """
        重置任务
        :return: None
        """
        self._task.reload()

    @property
    def isRuning(self):

        return self._run

    @property
    def task(self):

        return self._task

    @task.setter
    def task(self, newValue):

        if self.isRuning:
            raise OperationException("任务正在运行中，不允许重新注册")

        if isinstance(newValue, TaskIterator):
            self._task = newValue
        elif hasattr(newValue, "__call__"):
            self._task = TaskIterator(newValue, cycles=1)
        else:
            raise RegistryException("任务必须为可调用的,它可以是一个函数或TaskIterator实例对象")

class TaskManager:

    def __init__(self, result_q):

        self._result_q = result_q
        self._all_tasks = {}

    @classmethod
    def run_ever(cls, task_q, result_q):
        """
        接收任务并执行
        :param task_q: 任务队列
        :param output_q: 结果队列
        :return: None
        """
        manager = cls(result_q)

        while True:
            if task_q.empty():
                time.sleep(0.2)
                continue

            task_params = task_q.get()
            task_type = task_params[0]
            if task_type == "add":
                manager._add_task(*task_params[1:])
            elif task_type == "modify":
                manager._modify_task(*task_params[1:])
            elif task_type == "start":
                manager._start_task(*task_params[1:])
            elif task_type == "stop":
                manager._stop_task(*task_params[1:])
            elif task_type == "stop-all":
                manager._stop_all()

    def _add_task(self, task_name, task):
        """
        添加/注册任务
        :param task_name: 任务名称
        :param task: 任务实体
        :return: None
        """
        self._all_tasks[task_name] = Task(task_name, task, self._result_q)

    def _modify_task(self, task_name, task):
        """
        修改任务
        :param task_name: 任务名称
        :param task: 任务实体
        :return: None
        """
        self._all_tasks[task_name].task = task

    def _start_task(self, task_name, args, kwargs):
        """
        开启任务
        :param task_name: 任务名称
        :param args: 执行任务的位置参数
        :param kwargs: 执行任务的关键字参数
        :return: None
        """
        if self._all_tasks[task_name].isRuning:
            return

        self._all_tasks[task_name].reload()
        t = Thread(target=self._all_tasks[task_name].run, args=args, kwargs=kwargs)
        t.setDaemon(True)
        t.start()

    def _stop_task(self, task_name):
        """
        停止任务
        :param task_name: 任务名称
        :return: None
        """
        self._all_tasks[task_name].quit()

    def _stop_all(self):
        """
        停止所有任务
        :return: None
        """
        for task in self._all_tasks.values():
            task.quit()

pyqt_ipc/renderer.py
from PyQt5.QtWidgets import QApplication


__all__ = [
    "IPCRenderer"
]

class CallController:

    def __init__(self, callback):
        self._callbacks = [callback]

    def add(self, callback):
        """
        添加结果回调函数
        :param callback: 结果回调函数
        :return: None
        """
        self._callbacks.append(callback)

    def __call__(self, *args, **kwargs):
        """
        调用回调函数
        :param args:
        :param kwargs:
        :return:
        """
        for callback in self._callbacks:
            callback(*args, **kwargs)

        QApplication.processEvents()

class IPCRenderer:

    def __init__(self, ipcMain):

        self._ipcMain = ipcMain

    def on(self, task_name, callback):
        """
        添加任务监听
        :param task_name: 任务名称
        :param callback: 结果回调函数
        :return: None
        """
        if task_name in self._ipcMain.listen_always_tasks:
            self._ipcMain.listen_always_tasks[task_name].add(callback)
        else:
            new_always_tasks = self._ipcMain.listen_always_tasks
            new_always_tasks[task_name] = CallController(callback)
            self._ipcMain.listen_always_tasks = new_always_tasks

    def once(self, task_name, callback):
        """
        添加单次任务监听
        :param task_name: 任务名称
        :param callback: 结果回调函数
        :return: None
        """
        if task_name in self._ipcMain.listen_once_tasks:
            self._ipcMain.listen_once_tasks[task_name].add(callback)
        else:
            new_once_tasks = self._ipcMain.listen_once_tasks
            new_once_tasks[task_name] = CallController(callback)
            self._ipcMain.listen_once_tasks = new_once_tasks

example.py
import sys
import os
import time
from PyQt5.QtWidgets import QMainWindow, QApplication, QPushButton, QTextEdit
from pyqt_ipc import CreateIPC
from pyqt_ipc.task import TaskIterator

class Example(QMainWindow):

    def __init__(self):
        super(Example, self).__init__()

        self.initUi()

    def initUi(self):
        self.ipcMain, self.ipcReanderer = CreateIPC(self)

        self.setWindowTitle("PyQt-IPC Demo (by: zzmx-sudo)")
        self.resize(400, 400)
        self.btn = QPushButton("点我展示文本内容", self)
        self.btn.resize(self.btn.sizeHint())
        self.btn.move(10, 10)
        self.btn.clicked.connect(lambda :self.ipcMain.start("read_file", "./pyqt_ipc/demo_aaa.txt"))

        self.btn2 = QPushButton("取消文本展示", self)
        self.btn2.resize(self.btn.sizeHint())
        self.btn2.move(120, 10)
        self.btn2.clicked.connect(lambda: self.ipcMain.cancel("read_file"))

        self.textBox = QTextEdit(self)
        self.textBox.resize(380, 350)
        self.textBox.move(10, 40)

        self._setting_async_tasks()

        self.show()

    def _setting_async_tasks(self):
        self.ipcMain.registry("read_file", TaskIterator(self._read_file))
        self.ipcReanderer.on("read_file", self.textBox.append)
        # self.ipcReanderer.once("read_file", self.textBox.setText)

    @staticmethod
    def _read_file(file_path):
        time.sleep(1)

        if not os.path.exists(file_path):
            return ""

        with open(file_path, "r") as f:
            return f.read()


if __name__ == '__main__':
    app = QApplication(sys.argv)
    window = Example()
    sys.exit(app.exec_())